src/nvim/option.c

```diff
@@ -3731,69 +3598,148 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
+  }
+
+  if (value.type == kOptValTypeNumber) {
+    errmsg = validate_num_option((OptInt *)varp, &value.data.number);
+
+    // Don't change the value and return early if validation failed.
+    if (errmsg != NULL) {
+      return errmsg;
     }
-  } else if (flags & P_STRING) {
-    v.type = kOptValTypeString;
-    v.data.string.data = NULL;
   }
 
-  return v;
-}
+  // Set the new option value.
+  set_option_varp(varp, value);
+  // Remember where the option was set.
+  set_option_sctx_idx(opt_idx, opt_flags, current_sctx);
+  // May set global value for local option.
+  if ((opt_flags & (OPT_LOCAL | OPT_GLOBAL)) == 0) {
+    set_option_varp(get_varp_scope(opt, OPT_GLOBAL), value);
+  }
 
-static const char *set_option(int opt_idx, void *varp, OptVal *v, int opt_flags, char *errbuf,
-                              size_t errbuflen)
-{
-  const char *errmsg = NULL;
+  // Invoke the option specific callback function to validate and apply the new value.
+  bool doskip = false;
+  opt_did_set_cb_T did_set_cb;
 
-  bool value_checked = false;
+  if ((int *)varp == &p_force_on) {
+    did_set_cb = did_set_force_on;
+  } else if ((int *)varp == &p_force_off) {
+    did_set_cb = did_set_force_off;
+  } else {
+    did_set_cb = opt->opt_did_set_cb;
+  }
+  if (did_set_cb != NULL) {
+    // TODO(famiu): make os_oldval and os_newval use OptVal.
+    optset_T did_set_cb_args = {
+      .os_varp = varp,
+      .os_flags = opt_flags,
+      .os_oldval = old_value.data,
+      .os_newval = value.data,
+      .os_doskip = false,
+      .os_errbuf = NULL,
+      .os_errbuflen = 0,
+      .os_buf = curbuf,
+      .os_win = curwin
+    };
+
+    errmsg = did_set_cb(&did_set_cb_args);
+    doskip = did_set_cb_args.os_doskip;
+  }
+  if (doskip) {
+    return errmsg;
+  }
 
-  if (v->type == kOptValTypeBoolean) {
-    errmsg = set_bool_option(opt_idx, varp, (int)v->data.boolean, opt_flags);
-  } else if (v->type == kOptValTypeNumber) {
-    errmsg = set_num_option(opt_idx, varp, v->data.number, errbuf, errbuflen, opt_flags);
-  } else if (v->type == kOptValTypeString) {
-    errmsg = set_string_option(opt_idx, varp, v->data.string.data, opt_flags, &value_checked,
-                               errbuf, errbuflen);
+  // Check the bound for num options.
+  if (value.type == kOptValTypeNumber) {
+    errmsg
+      = check_num_option_bounds((OptInt *)varp, old_value.data.number, errbuf, errbuflen, errmsg);
   }
 
+  apply_optionset_autocmd(opt_idx, opt_flags,
+                          old_value,
+                          old_global_value,
+                          old_local_value,
+                          value,
+                          errmsg);
+
+  if (opt->flags & P_UI_OPTION) {
+    OptVal value_copy = optval_copy(optval_from_varp(value.type, varp));
+    ui_call_option_set(cstr_as_string(opt->fullname),
+                       optval_as_object(value_copy));
+  }
+
+  comp_col();                       // in case 'columns' or 'ls' changed
```

Comment:
this kind of indentation only makes sense if it's aligned with other comments.
```suggestion
  comp_col();  // in case 'columns' or 'ls' changed
```